### PR TITLE
[feature] Add support for all route params

### PR DESCRIPTION
**Pull request by vadolasi**

### CODE REVIEW

File: vue/router.py [hunk 1]
```diff
@@ -2,15 +2,28 @@
 from .factory import Wrapper, VueRouterFactory
 
 
+def init_dict(component):
```
Comment (stefanhoelzl) [inline vue/router.py:5]: why is this function needed?

**Comment by stefanhoelzl** [comment]
can you please also add unit and selenium tests for this?
the places would be here:
https://github.com/stefanhoelzl/vue.py/blob/master/tests/selenium/test_vuerouter.py
https://github.com/stefanhoelzl/vue.py/blob/master/tests/unit/test_factory/test_router.py

File: vue/router.py [hunk 2]
```diff
@@ -2,15 +2,28 @@
...
+
+
 class VueRouter(Wrapper):
     @classmethod
     def init_dict(cls):
         return VueRouterFactory.get_item(cls)
 
-    def __new__(cls):
-        return window.VueRouter.new(cls.init_dict())
+    def __new__(cls, VueRouterObject=window.VueRouter):
```
Comment (stefanhoelzl) [inline vue/router.py:14]: can you provide me some links to documentation how to use custom vue-router objects?
Just for me to understand how this works

File: vue/router.py [hunk 3]
```diff
@@ -2,15 +2,28 @@
...
+        window.Vue.use(VueRouterObject)
+        return VueRouterObject.new(cls.init_dict())
 
 
 class VueRoute:
-    def __new__(cls, path, component):
-        return {"path": path, "component": component}
+    def __new__(cls, path, component=None, components=None, **kwargs):
+        route = {"path": path}
```
Comment (stefanhoelzl) [inline vue/router.py:21]:
```suggestion
        route = {"path": path, **kwargs}
```

File: vue/router.py [hunk 4]
```diff
@@ -2,15 +2,28 @@
...
+        route.update(kwargs)
+        
+        if component is not None:
+            route["component"] = component.init_dict()
+        elif components is not None:
+            route["components"] = {k: init_dict(v) for k, v in components.items()}
```
Comment (stefanhoelzl) [inline vue/router.py:27]: please use clear variable names here:
```suggestion
            route["components"] = {
                name: component.init_dict() 
                for name, component in components.items()
            }
```

**Comment by stefanhoelzl** [comment]
just added some [contribution guidelines](https://github.com/stefanhoelzl/vue.py/blob/master/CONTRIBUTING.md).
Can you please update the commit messages accordingly?

File: vue/router.py (hunk 2 above)
Comment (vadolasi) [reply vue/router.py:14]: https://ionicframework.com/blog/announcing-the-ionic-vue-beta/

Comment (stefanhoelzl) [reply vue/router.py:14]: thanks, didn't now that there are different router implementations
can you stick with PEP-8 regarding naming? https://www.python.org/dev/peps/pep-0008/#function-and-variable-names
`VueRouterObject` -> `vue_router_class`
I would suggest to call it class, since it is not really a instance of the router if I get this correct?

File: vue/router.py [hunk 5]
```diff
@@ -2,15 +2,31 @@
...
+
 class VueRouter(Wrapper):
     @classmethod
     def init_dict(cls):
         return VueRouterFactory.get_item(cls)
 
-    def __new__(cls):
-        return window.VueRouter.new(cls.init_dict())
+    def __new__(cls, VueRouterObject=window.VueRouter):
+        window.Vue.use(VueRouterObject)
```
Comment (stefanhoelzl) [inline vue/router.py:15]: actually not sure about this here.
Maybe this should not be done implicit here but explicit in the application code?

File: vue/router.py [hunk 6]
```diff
@@ -2,15 +2,31 @@
...
+        return VueRouterObject.new(cls.init_dict())
 
 
 class VueRoute:
-    def __new__(cls, path, component):
-        return {"path": path, "component": component}
+    def __new__(cls, path, component=None, components=None, **kwargs):
+        route = {"path": path, **kwargs}
+        route.update(kwargs)
```
Comment (stefanhoelzl) [inline vue/router.py:22]: not needed anymore with the latest changes to the line above

**Comment by stefanhoelzl** [comment]
when you rebase your branch onto my current master, then the merge conflict should be resolved.
I hope then TravisCI will pick up this PR and run all the tests

File: vue/router.py (hunk 5 above)
Comment (vadolasi) [reply vue/router.py:15]: I find it convenient even without using a different class.

**Comment by vadolasi** [comment]
Nested routes are not working properly

File: vue/router.py (hunk 5 above)
Comment (stefanhoelzl) [reply vue/router.py]: Thanks for changing it!

I agree, it is convenient.
But there are some downsides, which I think are not worth the gain in convenience:
* It introduces a different behavior than vue.js
* It is a side-effect which you do not expect when you do not know the class
* It introduces a new dependency to `window.Vue`

**Comment by stefanhoelzl** [comment]
> Nested routes are not working properly

are some tests failing?
see my comment above how to get TravisCI to test the PR, then I can see why it fails, and maybe help to solve it

**Comment by vadolasi** [comment]
When everything is right I will create another cleaner PR that follows the contribution guidelines. And sorry for the mistakes, I am only 15 years old and this is the first time I have contributed to an open source project. In addition, I live in Brazil and do not understand English.

**Comment by stefanhoelzl** [comment]
> When everything is right I will create another cleaner PR that follows the contribution guidelines. And sorry for the mistakes, I am only 15 years old and this is the first time I have contributed to an open source project. In addition, I live in Brazil and do not understand English.

No worries, we will get this PR ready to be merged 😄 

It is great that you want to contribute to some project, nothing to apologize there!
I will try to guide you as good as possible if you like!

Unfortunately I do not speak any Portuguese, so we have to stick with englisch 😄

**Comment by stefanhoelzl** [comment]
finally I got the TravisCI to pick up the PR correctly!

You will see that a check still fails
https://travis-ci.com/github/stefanhoelzl/vue.py/builds/172236655

These are just some formatting errors, by running
`black .`
they should be resolved

[black](https://github.com/psf/black) is a auto-formatting tool, so that all python files are formatted the same.

when you set-up everything with `make env.up` as described in the README, than this should already be installed

**Comment by stefanhoelzl** [comment]
btw. the tests you are added are looking great. Thanks a lot!

File: vue/router.py [hunk 7]
```diff
@@ -7,10 +7,20 @@ class VueRouter(Wrapper):
     def init_dict(cls):
         return VueRouterFactory.get_item(cls)
 
-    def __new__(cls):
-        return window.VueRouter.new(cls.init_dict())
+    def __new__(cls, vue_router_class=window.VueRouter):
```
Comment (stefanhoelzl) [inline vue/router.py:10]: This is the last request for a change I have:
Can we revert this change, and not have it in this PR?

I like this feature and would like to have it, but lets keep the scope of this PR only to adding support for the rest of the route params.

I would propose adding this in a second PR, and also take care of proper tests of this feature in the second PR

File: tests/unit/test_factory/test_router.py [hunk 8]
```diff
@@ -6,6 +6,37 @@ def test_path_and_component(self):
...
+                "default": VueComponent.init_dict(),
+                "named": VueComponent.init_dict(),
+            },
+        }
+
+    def test_path_and_components_and_children(self):
+        route = VueRoute(
+            "/path",
+            VueComponent,
+            children=[VueRoute("path", VueComponent), VueRoute("path2", VueComponent),],
```
Comment (stefanhoelzl) [inline tests/unit/test_factory/test_router.py:25]:
```suggestion
            children=[VueRoute("/path", VueComponent), VueRoute("/path2", VueComponent),],
```
this should fix the failing tests.

File: tests/unit/test_factory/test_router.py [hunk 9]
```diff
@@ -6,6 +6,37 @@ def test_path_and_component(self):
...
+                "default": VueComponent.init_dict(),
+                "named": VueComponent.init_dict(),
+            },
+        }
+
+    def test_path_and_components_and_children(self):
+        route = VueRoute(
+            "/path",
+            VueComponent,
+            children=[VueRoute("/path", VueComponent), VueRoute("/path2", VueComponent),],
```
Comment (stefanhoelzl) [inline tests/unit/test_factory/test_router.py:25]:
```suggestion
            children=[
                VueRoute("/path", VueComponent),
                VueRoute("/path2", VueComponent),
            ],
```
fixes formatting issues

**Comment by stefanhoelzl** [comment]
congrats everything in your first PR is green 💚  grab a Pão de queijo and celebrate it 🥳

**Comment by vadolasi** [comment]
I didn't understand "Pão de Queijo", but that's ok kkkkk (that's how we "write laughs" in Portuguese). Thanks for the help!
